pkg/pmem-csi-driver/controllerserver-node.go

```diff
@@ -74,6 +74,21 @@ func NewNodeControllerServer(nodeID string, dm pmdmanager.PmemDeviceManager, sm
 		}
 
 		for _, id := range ids {
+			// retrieve volume info
+			vol := &nodeVolume{}
+			if err := sm.Get(id, vol); err != nil {
+				klog.Warningf("Failed to retrieve volume info for id %q from state: %v", id, err)
+				continue
+			}
+
+			v, err := parameters.Parse(parameters.NodeVolumeOrigin, vol.Params)
+			if err != nil {
+				klog.Warningf("Failed to parse volume parameters for volume %q: %v", id, err)
+			}
+			if v.GetDeviceMode() != dm.GetName() {
+				continue
```

Comment:
I wonder whether we should really ignore the volume. For `NodePublishVolume` that's kind of okay (results in "not found" error, right?), albeit a bit confusing (the volume exists, it's just not usable at the moment).

But what about a repeated `CreateVolume` call (can happen when our initial response didn't get through to Kubernetes)? Is that still idempotent or will we end up with one volume created in direct mode and another in LVM mode for the same Kubernetes PVC?

Reply:
> But what about a repeated `CreateVolume` call (can happen when our initial response didn't get through to Kubernetes)? Is that still idempotent or will we end up with one volume created in direct mode and another in LVM mode for the same Kubernetes PVC?

I believe, If we want to cover theses cases we should fully support volumes in multiple modes, not just that driver switching modes.

Reply:
If that's what it takes to prevent bad things from happening when admins make mistakes, then yes, we'll have to do that.

Reply:
> If that's what it takes to prevent bad things from happening when admins make mistakes, then yes, we'll have to do that.

This is a big design change, PMEM-CSI driver was not designed to support volumes with multiple device modes at a time.  And, I doubt how far the current code helps to switch modes. Say, if the driver deployed in `LVM` mode and then redeploying with `ndctl` mode is really useful as the space used for "lvm" could not be unitized by driver anymore. 

The intention of this change is: not to remove the volumes that were created in the previous mode from the state. That gives the admin to revert to the previous mode and still could access those volumes.